### ISSUE TITLE
Uses of Consumer, Producer, Function, Predicate Interfaces

### ISSUE THREAD

**Issue by ongyongen** [nus-cs2030/2122-s2#252]
# Description

Hi! Can I clarify the use cases of Consumer, Producer, Function, Predicate interfaces?
Is it because when we create our own ImList, it cannot use stream() directly so we cannot directly apply filter, map etc (like what we can do with Java's List). So we create methods based on these interfaces to allow us to use map, filter etc?

Thanks!

**Comment by jamietan123** [comment]
^ you can use these methods like filter, map etc directly on ImList

**Comment by celine-chung** [comment]
Thinking about this more intuitively/ in layman terms, we perform functions like map and filter on lists of values. These functions iterate through the list to do the behaviour you want. A stream is one sort of list you can perform functions on and these methods are defined in the java API while ImList is another sort of list you can perform functions on and these methods were implemented by prof in the lectures. 

The consumer, producer, function and predicate interfaces are what we need to input into the functions to pass our intended behaviour to them when we use the function. For instance map: we need to tell ImList how exactly we want to modify each value in the list (i.e. the behaviour we pass on to map). We input this behaviour as the argument into the map method using the Function interface! (If you want to check what interface you need to implement in order to use a particular function, just check the stream API/ implementation of ImList) Hope my understanding helps:-)
